### DISCUSSION TITLE
external hook in NextJS, document is not defined

### DISCUSSION THREAD

**Discussion by huypham1411** [vercel/next.js#52859] [Help] [answered]
### Summary

My current project needs this hook to get the barcode scan data from an external device but NextJs show this error 
- error node_modules/use-scan-detection-react18/dist/index.js (9:0) @ useScanDetection
- error ReferenceError: document is not defined

I wonder if any way to resolve this issue, appreciate it if anyone can help me. Thanks!

### Additional information

```js
https://www.npmjs.com/package/use-scan-detection-react18?activeTab=readme
```

**Comment by masterbater** [comment]
If you are using App Router add client directive
```typescript
'use client' //Add this
import React, { useState } from 'react';
import useScanDetection from 'use-scan-detection';

const Input = () => {
    const [value, setValue] = useState("");

    useScanDetection({
        onComplete: setValue,
        minLength: 13 // EAN13
    });

    return (
        <input 
            value={value} 
            type="text"
        />
    );
};

export default Input
```
If you are not,  detect window if undefined or use useEffect

**Reply by huypham1411** [reply]
My code has already got 'use client' but still does not work so I try to find help

**Comment by icyJoseph** [comment]
Hi,

Adding `use client` doesn't make a component run only client side. That's, sadly, a spread misconception. 

Components with `use client` are still rendered on the server and client, just like they always have. Server components are the new thing, and they render only server side.

So you need to guard the usage of `useScanDetection` for example:

```tsx
'use client'
import { useState, useEffect } from 'react';
import useScanDetection from 'use-scan-detection';

const InputScan = () => {
    const [value, setValue] = useState("");

    useScanDetection({
        onComplete: setValue,
        minLength: 13 
    });

    return (
        <input 
            value={value} 
            type="text"
        />
    );
};

export const Input = () => { 
  const [isServer,setServer] = useState(true);
  useEffect(() => { setServer(false) } ,[]);

  if (isServer) return null;

  return <InputScan />
}
```

In this particular case, the issue is that `useScanDetection` takes a container as prop, which is defaulted to `document`, so we want to avoid that line during SSR.

**Reply by huypham1411** [reply, marked as answer]
so I found another way to fix this is by putting the useScanDetection hook in one component and then dynamic import it to the page instead using it directly on the page, since I need to use an external scan repeatedly to import more data when I scan

**Reply by icyJoseph** [reply]
Yeah that's also a valid approach.

**Reply by icyJoseph** [reply]
I perhaps should complement that, lazy/dynamic loading is not necessarily a great idea. There's some wrapper code that goes around the dynamically loaded component, and it also means an extra request to the server, yada yada. 

Code splitting is a great technique though! It is just a trade-off that we have to be aware of.
